### PR TITLE
feat: Allow multiple data files as input

### PR DESCRIPTION
**Pull request by lalo**
implements #1895

- [x] happy path
- [ ] cache debug

### CODE REVIEW

**Review by olgavrou — APPROVED** [review]

File: test/core.vwtest.json [hunk 1]
```diff
@@ -5123,5 +5123,35 @@
...
+    },
+    "input_files": [
+      "train-sets/dsjson_cb_part1.json",
+      "train-sets/dsjson_cb_part2.json"
+    ]
+  },
+  {
+    "id": 400,
+    "desc": "test multi d positional -- based on test 316 to share same metrics of the complete file dsjson_cb.json (which is split into two here)",
+    "vw_command": "--dsjson --cb_explore_adf --epsilon 0.2 --quadratic GT --extra_metrics metrics_time.json train-sets/dsjson_cb_part1.json train-sets/dsjson_cb_part2.json",
```
Comment (ataymano) [inline test/core.vwtest.json:5144]: do we want to allow multiple positional arguments for input file? even single one is constant source of problems when someone accidentally adding wrong options. Multiple positional arguments could be source of problems that could be hard to notice (like every wrong commandline option will be treated as input file?). Maybe we should explicitly deprecate that?

File: vowpalwabbit/core/src/parse_args.cc [hunk 2]
```diff
@@ -419,13 +419,7 @@ input_options parse_source(VW::workspace& all, options_i& options)
   // Check if the options provider has any positional args. Only really makes sense for command line, others just return
   // an empty list.
   const auto positional_tokens = options.get_positional_tokens();
-  if (positional_tokens.size() == 1) { all.data_filename = positional_tokens[0]; }
-  else if (positional_tokens.size() > 1)
-  {
-    all.logger.err_warn(
```
Comment (lalo) [inline vowpalwabbit/core/src/parse_args.cc:425]: I'm reverting this warning too, so maybe we should keep it?

File: test/train-sets/ref/rcv1_raw_cb_explore_pass2_split.stderr [hunk 3]
```diff
@@ -0,0 +1,35 @@
+predictions = rcv1_raw_cb_explore.preds
+creating cache_file = train-sets/rcv1_raw_cb_small_p2.vw.merged.cache
```
Comment (lalo) [inline test/train-sets/ref/rcv1_raw_cb_explore_pass2_split.stderr:2]: gotcha: one cache per run

File: test/core.vwtest.json (hunk 1 above)
Comment (jackgerrits) [reply test/core.vwtest.json:5144]: 100% agree. We should not be expanding the positional input.

Comment (ataymano) [reply test/core.vwtest.json:5144]: And probably the same about mixed mode (like one input is positional, another - explicit)?

Comment (lalo) [reply test/core.vwtest.json:5144]: yes, but the log should provide help to decipher it. See test 404 for a weird mixed mode side effect in ordering.

File: python/vowpalwabbit/pyvw.py [hunk 4]
```diff
@@ -176,9 +176,11 @@ def __str__(self):
             if self.is_flag():
                 return "--{}".format(self.name)
             else:
-                # missing list case
                 if isinstance(self.value, list):
-                    return "**NOT_IMPL**"
+                    res = ""
+                    for ele in self.value:
+                        res += f"--{self.name} {ele} "
```
[the remark this answers is not in the document]
Comment (lalo) [reply python/vowpalwabbit/pyvw.py]: outdated diff but might still apply

File: cs/cli/vw_arguments.h [hunk 5]
```diff
@@ -36,11 +36,16 @@ public ref class VowpalWabbitArguments
   float m_power_t;
 
   internal : VowpalWabbitArguments(VW::workspace* vw)
```
Comment (jackgerrits) [inline cs/cli/vw_arguments.h:38]: Does this mean C# cannot support multiple data files?

File: cs/cli/vw_arguments.h [hunk 6]
```diff
@@ -36,11 +36,16 @@ public ref class VowpalWabbitArguments
...
-      : m_data(gcnew String(vw->data_filename.c_str()))
-      , m_finalRegressor(gcnew String(vw->final_regressor_name.c_str()))
+      : m_finalRegressor(gcnew String(vw->final_regressor_name.c_str()))
       , m_testonly(!vw->training)
       , m_passes((int)vw->numpasses)
   {
+    if (vw->data_filenames.size() >= 1) { m_data = gcnew String(vw->data_filenames.front().c_str()); }
+    else
+    {
+      m_data = gcnew String("");
```
Comment (jackgerrits) [inline cs/cli/vw_arguments.h:46]: Please throw rather than silently ignore multiple test files

File: cs/cli/vw_arguments.h (hunk 5 above)
Comment (lalo) [reply cs/cli/vw_arguments.h:38]: yes, CLI for now and python thanks to the duck typing and support for lists (-q and other args)

File: cs/cli/vw_arguments.h (hunk 6 above)
Comment (lalo) [reply cs/cli/vw_arguments.h:46]: added throw just in case, but not entirely sure it can reach that state.

**Comment by lalo** [comment]
https://github.com/VowpalWabbit/vowpal_wabbit/blob/c9cb804a74c94df4a00756c78f6e204526fcb7a9/vowpalwabbit/core/src/parser.cc#L311

File: test/core.vwtest.json (hunk 1 above)
Comment (ataymano) [reply test/core.vwtest.json:5144]: hm, it still seems too scary to be honest - basically we are introducing a lot of ambiguity in parsing - for example if you have file with the same name as valid command line argument: let's say you have files with names "-q" and "sa" - then what command line "--cb_explore_adf -q sa -d file.txt" is supposed to do?
I understand that it is extreme corner case, but that change is increasing their amount substantially.

File: vowpalwabbit/core/src/parse_args.cc [hunk 7]
```diff
@@ -419,27 +421,58 @@ input_options parse_source(VW::workspace& all, options_i& options)
...
-        "Multiple data files passed as positional parameters, only the first one will be "
-        "read and the rest will be ignored.");
-  }
 
-  if (parsed_options.daemon || options.was_supplied("pid_file") || (options.was_supplied("port") && !all.active))
+  if (!positional_tokens.empty() && !all.data_filenames.empty())
+  { THROW("You cannot mix --data/-d argument with positional arguments. Pick one and use consistently."); }
```
Comment (lalo) [inline vowpalwabbit/core/src/parse_args.cc:426]: added this throw to allow users to pick either all --data or positional, would this be helpful as a previous step to deprecating positional?

**Pull request closed without merging**